sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -195,14 +195,35 @@ abstract class Expression extends TreeNode[Expression] {
   }
 
   /**
-   * Returns true when two expressions will always compute the same result, even if they differ
+   * Returns true when two expressions will always compute the same output, even if they differ
    * cosmetically (i.e. capitalization of names in attributes may be different).
    *
    * See [[Canonicalize]] for more details.
+   *
+   * This method should be used (instead of `sameResult`) when comparing if 2 expressions are the
+   * same and one can replace the other (eg. in Optimizer/Analyzer rules where we want to replace
+   * equivalent expressions). It should not be used (and `sameResult` should be used instead) when
+   * comparing if 2 expressions produce the same results (in this case `semanticEquals` can be too
+   * strict).
    */
   def semanticEquals(other: Expression): Boolean =
     deterministic && other.deterministic && canonicalized == other.canonicalized
 
+  /**
+   * Returns true when two expressions will always compute the same result, even if the output may
+   * be different, because of different names or similar differences.
```

Comment:
I think here `output` is a bit confusing. Do we mean the output names?

Reply:
So sameResult returns if the evaluated results between two expressions are exactly the same?

Reply:
yes, I mean: `sameResult` returns true if 2 expressions return the same data even though from plan perspective they are not the same (eg. the output name/exprIds is different as in this case), while `semanticEquals` ensure they are the same from plan perspective too. If you have better suggestions how to rephrase this, I am happy to improve it. Thanks.

Reply:
How about replace `output` with `output from plan perspective`?

```
Returns true when two expressions will always compute the same result, even if the output
from plan perspective may be different, because of different names or similar differences.
```